From: Sinthu Raja <sinthu.raja@ti.com>

The OV5640 camera sensor is interfaced with CSI2RX through a MIPI header.
The sensor shall be configured through the I2C which is interfaced with
the camera sensor via an I2C switch.

MAIN_I2C ===> I2C_switch(TCA954) ====> \\
                                         ==>MIPI header ===> OV5640
                         CSI2RX0 ====> //

Add node for the I2C switch through which the camera sensor is connected.

Add node for OV5640 camera connected to port 0 of the CSI bridge.

Add pinmux for the reset gpio and CSI2 Reference clock to the camera
sensor.

Signed-off-by: Sinthu Raja <sinthu.raja@ti.com>
---
 arch/arm64/boot/dts/ti/Makefile               |  1 +
 .../boot/dts/ti/k3-j721e-eaik-csi2-ov5640.dts | 89 +++++++++++++++++++
 2 files changed, 90 insertions(+)
 create mode 100644 arch/arm64/boot/dts/ti/k3-j721e-eaik-csi2-ov5640.dts

diff --git a/arch/arm64/boot/dts/ti/Makefile b/arch/arm64/boot/dts/ti/Makefile
index 48b6372a9119..26dd75ace661 100644
--- a/arch/arm64/boot/dts/ti/Makefile
+++ b/arch/arm64/boot/dts/ti/Makefile
@@ -25,6 +25,7 @@ dtb-$(CONFIG_ARCH_K3) += k3-j721e-common-proc-board-infotainment.dtbo
 dtb-$(CONFIG_ARCH_K3) += k3-j721e-eaik.dtb
 dtb-$(CONFIG_ARCH_K3) += k3-j721e-edgeai-apps.dtbo
 dtb-$(CONFIG_ARCH_K3) += k3-j721e-vision-apps.dtbo
+dtb-$(CONFIG_ARCH_K3) += k3-j721e-eaik-csi2-ov5640.dtbo
 
 dtb-$(CONFIG_ARCH_K3) += k3-j7200-common-proc-board.dtb
 
diff --git a/arch/arm64/boot/dts/ti/k3-j721e-eaik-csi2-ov5640.dts b/arch/arm64/boot/dts/ti/k3-j721e-eaik-csi2-ov5640.dts
new file mode 100644
index 000000000000..d8dac41787ef
--- /dev/null
+++ b/arch/arm64/boot/dts/ti/k3-j721e-eaik-csi2-ov5640.dts
@@ -0,0 +1,89 @@
+// SPDX-License-Identifier: GPL-2.0
+/**
+ * DT Overlay for OV5640 Camera on I2C bus interfaced to CSI2 with J721E-EAIK board.
+ *
+ * Copyright (C) 2021 Texas Instruments Incorporated - https://www.ti.com/
+ */
+
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/interrupt-controller/irq.h>
+#include <dt-bindings/pinctrl/k3.h>
+
+&main_pmx0 {
+	csi2_exp_reset_pins_default: csi2-exp-reset-pins-default {
+		pinctrl-single,pins = <
+			J721E_IOPAD(0x140, PIN_OUTPUT, 7) /* (AG29) PRG0_PRU1_GPO16.GPIO0_79 */
+		>;
+	};
+
+	csi2_exp_refclk_pins_default: csi2-exp-refclk-pins-default {
+		pinctrl-single,pins = <
+			J721E_IOPAD(0x1A4, PIN_OUTPUT, 3) /* (W26) RGMII6_RXC.AUDIO_EXT_REFCLK2 */
+		>;
+	};
+};
+
+&k3_clks {
+	/* Confiure AUDIO_EXT_REFCLK2 pin as output */
+	pinctrl-names = "default";
+	pinctrl-0 = <&csi2_exp_refclk_pins_default>;
+};
+
+&main_i2c3 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+
+	i2c-switch@70 {
+		compatible = "nxp,pca9543";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		reg = <0x70>;
+
+		/* CAM0 I2C */
+		ti_cam0_i2c: i2c@0 {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0>;
+
+			ov5640: camera@3c {
+				compatible = "ovti,ov5640";
+				reg = <0x3c>;
+
+				pinctrl-names = "default";
+				pinctrl-0 = <&csi2_exp_reset_pins_default>;
+				reset-gpios = <&main_gpio0 79 GPIO_ACTIVE_LOW>;
+
+				/* C_AUDIO_REFCLK3 -> RGMII6_RXC (W26) */
+				clocks = <&k3_clks 157 371>;
+				clock-names = "xclk";
+
+				/* HSDIV3_16FFT_MAIN_4_HSDIVOUT2_CLK -> REFCLK2 */
+				assigned-clocks = <&k3_clks 157 371>;
+				assigned-clock-parents = <&k3_clks 157 400>;
+				assigned-clock-rates = <25000000>;
+
+				port {
+					csi2_cam0: endpoint {
+						remote-endpoint = <&csi2rx0_in_sensor>;
+						clock-lanes = <0>;
+						data-lanes = <1 2>;
+					};
+				};
+			};
+		};
+
+	};
+};
+
+&csi0_port0 {
+	csi2rx0_in_sensor: endpoint {
+		remote-endpoint = <&csi2_cam0>;
+		bus-type = <4>; /* CSI2 DPHY. */
+		clock-lanes = <0>;
+		data-lanes = <1 2>;
+	};
+
+};
